### PR TITLE
Fixed mag fusion bugs in EK2 and EK3

### PR DESCRIPTION
**Pull request by tridge**
This fixes several bugs in both EK2 and EK3 mag fusion

bug1: fixed 3D mag fusion in EK3 for a 2nd flight. We were previously never doing 3D fusion on copters if we had previously stopped fusion due to landing
for example, before fix:
image
after fix:
image

bug2: fixed reset of body variances when we start 3D fusion in both EK2 and EK3

when we switch from mag fusion inhibited to being enabled we need to reset the variances. This matters for when you land and change payload (or change battery). We had very slow learning of offsets on the 2nd flight as the variances had been reset to zero.
for example, with EK2 with a change in the simulated compass of 100mGauss on first takeoff, then change back for 2nd takeoff:
image
the above shows fast learning for 1st takeoff, slow for 2nd
with the fix we get fast learning with both flights
image

this PR also fixes SITL to enable the compass scale parameters by default, so that we use the table based EKF mag earth field init, which is a better match for real usage

This also fixes the mag variance reset when we get an in-flight compass change

### CODE REVIEW

File: libraries/AP_NavEKF3/AP_NavEKF3_Control.cpp [hunk 1]
```diff
@@ -91,6 +91,12 @@ void NavEKF3_core::setWindMagStateLearningMode()
         manoeuvring = false;
     }
 
+    if (finalInflightYawInit && !finalInflightMagInit) {
```
Comment (priseborough) [inline libraries/AP_NavEKF3/AP_NavEKF3_Control.cpp:94]: Better to fix the root cause which I understand to be the failure to set magFieldLearned to false when on ground. When I reverted this patch and added one to reset magFieldLearned when on ground, the second takeoff was doing mag fusion.

File: libraries/AP_NavEKF3/AP_NavEKF3_Control.cpp [hunk 2]
```diff
@@ -114,6 +114,8 @@ void NavEKF3_core::setWindMagStateLearningMode()
     if (!inhibitMagStates && setMagInhibit) {
         inhibitMagStates = true;
         updateStateIndexLim();
+        // we need to learn variances from scratch
```
Comment (priseborough) [inline libraries/AP_NavEKF3/AP_NavEKF3_Control.cpp:117]: We shouldn't have to do this here if we are doing the reset in CovariancePrediction.

File: libraries/AP_NavEKF2/AP_NavEKF2_Control.cpp [hunk 3]
```diff
@@ -105,6 +105,7 @@ void NavEKF2_core::setWindMagStateLearningMode()
     bool setMagInhibit = !magCalRequested || magCalDenied;
     if (!inhibitMagStates && setMagInhibit) {
         inhibitMagStates = true;
+        resetMagBodyVariances();
```
Comment (priseborough) [inline libraries/AP_NavEKF2/AP_NavEKF2_Control.cpp:108]: We shouldn't have to do this here is we are doing a reset in the CovariancePrediction.

**Comment by priseborough** [comment]
See https://github.com/tridge/ardupilot/pull/33

**Comment by tridge** [comment]
@priseborough thanks for the feedback, I have re-worked as we discussed and re-tested

**Comment by tridge** [comment]
I flew this yesterday on a plane with no issues
http://uav.tridgell.net/EKF/log32.bin

**Review by priseborough — APPROVED** [review]

**Pull request merged**